server/jdkhttp-server/src/main/scala/sttp/tapir/server/jdkhttp/internal/ParsedMultiPart.scala

```diff
@@ -33,59 +51,153 @@ case class ParsedMultiPart(headers: Map[String, Seq[String]], body: Array[Byte])
+    }
+
+    private def parseInitialBoundary(currentByte: Int): Unit = {
+      val foundBoundary = lookForBoundary(currentByte, boundaryBytes)
+      if (foundBoundary) {
+        changeState(LookForEndOrNewLine)
+      }
+    }
+
+    private def parsePartHeaders(): Unit = {
+      if (circularBuffer.endsWith(CRLF)) {
+        circularBuffer.getString match {
+          case "--\r\n"                          => done = true
+          case headerLine if !headerLine.isBlank => addHeader(headerLine)
+          case _                                 => changeState(ParsePartBody)
+        }
+        circularBuffer.reset()
+      } else if (circularBuffer.isFull) {
+        throw new RuntimeException(
+          s"Reached max size of $bufferSize bytes before reaching newline when parsing header."
         )
-      case None => changeState(AfterBoundary)
+      }
     }
-  }
 
-  def parseMultipartBody(httpExchange: HttpExchange, boundary: String): Seq[ParsedMultiPart] = {
-    val reader = new BufferedReader(new InputStreamReader(httpExchange.getRequestBody))
-    val initialParseState: ParseData = ParseData(empty, List.empty, Default)
-    Iterator
-      .continually(reader.readLine())
-      .takeWhile(_ != null)
-      .foldLeft(initialParseState) { case (state, line) =>
-        state.parseState match {
-          case Default if line.startsWith(boundary)           => state.changeState(AfterBoundary)
-          case Default                                        => state
-          case AfterBoundary if line.trim.isEmpty             => state.changeState(AfterHeaderSpace)
-          case AfterBoundary                                  => state.addHeader(line)
-          case AfterHeaderSpace if !line.startsWith(boundary) => state.addBody(line.getBytes())
-          case AfterHeaderSpace                               => state.completePart()
+    private def parsePartBody(currentByte: Int): Unit = {
+      bodySize += 1
+      val foundBoundary = lookForBoundary(currentByte, prefixedBoundaryBytes)
+      convertStreamToFileIfThresholdMet()
+      if (foundBoundary) {
+        val bytes = circularBuffer.getBytes
+        val bytesToWrite = bytes.slice(0, bytes.length - prefixedBoundaryBytes.length)
+        bytesToWrite.foreach(byte => stream.underlying.write(byte.toInt))
+        val bodyInputStream = stream match {
+          case FileStream(tempFile, stream) => stream.close(); new FileInputStream(tempFile)
+          case ByteStream(stream)           => new ByteArrayInputStream(stream.toByteArray)
         }
+        completePart(bodyInputStream)
+        stream = ByteStream()
+        bodySize = 0
+      } else if (numMatchedBoundaryChars == 0) {
+        circularBuffer.getBytes.foreach(byte => stream.underlying.write(byte.toInt))
+        circularBuffer.reset()
+      }
+    }
+
+    private def lookForBoundary(currentByte: Int, boundary: Array[Byte]): Boolean = {
+      if (currentByte == boundary(numMatchedBoundaryChars)) {
```

Comment:
this might be problematic if the boundary has some common substring with the body. E.g. let's say the boundary is `AAB...`

The body goes: `AAAB...`, so we have one `A` byte from the "proper" body, and then the boundary. We read `A`, `A`, advancing `numMatchedBoundaryChars`, but then we get a mismatch, which resets the counter. 

So now we'll continue on reading `AB...`, and we'll miss the boundary.

Instead, we should have retreated our steps and checked if some other prefix didn't match. I think it's a well known problem in CS ... maybe https://en.wikipedia.org/wiki/Knuth–Morris–Pratt_algorithm ?

Either way, we'll need tests for this :)

Reply:
Thats true! Good catch :+1: 
I think I'll get KMP working and try to add some tests.